File: backend/models/distribution.py
"""
Phase 10 — Multi-Platform Distribution models.

Persists connected platform accounts (encrypted tokens), the scheduled-post
queue, and a record of what was actually published. Adapted from the original
SQLite plan (docs/archive/planning/PHASE_10_DISTRIBUTION_PLAN.md) to the current
Postgres/SQLAlchemy + per-user multi-tenant model.
"""

from sqlalchemy import (
    Boolean,
    Column,
    DateTime,
    ForeignKey,
    Integer,
    String,
    Text,
    UniqueConstraint,
)
from sqlalchemy.sql import func

from backend.database import Base

# Platforms the distribution layer knows about.
SUPPORTED_PLATFORMS = (
    "linkedin",
    "twitter",
    "facebook",
    "instagram",
    "tiktok",
    "youtube",
    "stub",  # dry-run / demo target — never touches a real network
)


class PlatformCredential(Base):
    """A connected social account for a user (optionally scoped to a client)."""

    __tablename__ = "platform_credentials"
    __table_args__ = (
        UniqueConstraint("user_id", "client_id", "platform", name="uq_platform_credential"),
        {"extend_existing": True},
    )

    id = Column(String, primary_key=True)
    user_id = Column(String, ForeignKey("users.id"), nullable=False, index=True)
    client_id = Column(String, ForeignKey("clients.id"), nullable=True, index=True)
    platform = Column(String, nullable=False)
    # Fernet-encrypted at the service layer; never store plaintext tokens.
    access_token = Column(Text, nullable=False)
    refresh_token = Column(Text, nullable=True)
    token_expires_at = Column(DateTime(timezone=True), nullable=True)
    # Facebook Page id / Instagram business account id / LinkedIn org id, etc.
    account_ref = Column(String, nullable=True)
    display_name = Column(String, nullable=True)
    is_active = Column(Boolean, default=True, nullable=False)
    created_at = Column(DateTime(timezone=True), server_default=func.now())
    updated_at = Column(DateTime(timezone=True), onupdate=func.now())


class ScheduledPost(Base):
    """A post queued for publishing to one platform (the post_queue)."""

    __tablename__ = "scheduled_posts"
    __table_args__ = ({"extend_existing": True},)

    id = Column(String, primary_key=True)
    user_id = Column(String, ForeignKey("users.id"), nullable=False, index=True)
    project_id = Column(String, ForeignKey("projects.id"), nullable=True, index=True)
    client_id = Column(String, ForeignKey("clients.id"), nullable=True, index=True)
    post_id = Column(String, ForeignKey("posts.id"), nullable=True)
    platform = Column(String, nullable=False)
    content = Column(Text, nullable=False)
    media_url = Column(String, nullable=True)
    scheduled_for = Column(DateTime(timezone=True), nullable=False, index=True)
    # pending | posting | posted | failed | canceled
    status = Column(String, nullable=False, default="pending", index=True)
    posted_at = Column(DateTime(timezone=True), nullable=True)
    platform_post_id = Column(String, nullable=True)
    platform_url = Column(String, nullable=True)
    error_message = Column(Text, nullable=True)
    retry_count = Column(Integer, nullable=False, default=0)
    created_at = Column(DateTime(timezone=True), server_default=func.now())


class PostedContent(Base):
    """An immutable record of a successful publish (for analytics + dedup)."""

    __tablename__ = "posted_content"
    __table_args__ = ({"extend_existing": True},)

    id = Column(String, primary_key=True)
    user_id = Column(String, ForeignKey("users.id"), nullable=False, index=True)
    scheduled_post_id = Column(String, ForeignKey("scheduled_posts.id"), nullable=False, index=True)
    platform = Column(String, nullable=False)
    platform_post_id = Column(String, nullable=False)
    platform_url = Column(String, nullable=True)
    content_hash = Column(String, nullable=True, index=True)
    posted_at = Column(DateTime(timezone=True), server_default=func.now())

File: operator-dashboard/src/api/distribution.ts
import apiClient from './client';

export interface PlatformCredential {
  id: string;
  platform: string;
  client_id?: string | null;
  account_ref?: string | null;
  display_name?: string | null;
  is_active: boolean;
}

export interface ScheduledPost {
  id: string;
  platform: string;
  content: string;
  status: string;
  scheduled_for: string;
  posted_at?: string | null;
  platform_url?: string | null;
  error_message?: string | null;
  retry_count: number;
}

export interface OAuthStatus {
  configured: string[];
  all: string[];
}

export const distributionApi = {
  async platforms(): Promise<string[]> {
    const { data } = await apiClient.get<{ platforms: string[] }>('/api/distribution/platforms');
    return data.platforms;
  },
  async oauthStatus(): Promise<OAuthStatus> {
    const { data } = await apiClient.get<OAuthStatus>('/api/distribution/oauth/status');
    return data;
  },
  async oauthStart(platform: string): Promise<string> {
    const { data } = await apiClient.get<{ authorize_url: string }>(
      `/api/distribution/oauth/${platform}/start`,
    );
    return data.authorize_url;
  },
  async patchCredential(
    id: string,
    body: { account_ref?: string; display_name?: string; is_active?: boolean },
  ): Promise<PlatformCredential> {
    const { data } = await apiClient.patch<PlatformCredential>(
      `/api/distribution/credentials/${id}`,
      body,
    );
    return data;
  },
  async listCredentials(): Promise<PlatformCredential[]> {
    const { data } = await apiClient.get<PlatformCredential[]>('/api/distribution/credentials');
    return data;
  },
  async connect(body: {
    platform: string;
    access_token: string;
    display_name?: string;
    account_ref?: string;
    client_id?: string;
  }): Promise<PlatformCredential> {
    const { data } = await apiClient.post<PlatformCredential>('/api/distribution/credentials', body);
    return data;
  },
  async deleteCredential(id: string): Promise<void> {
    await apiClient.delete(`/api/distribution/credentials/${id}`);
  },
  async schedule(body: {
    platform: string;
    content: string;
    scheduled_for: string;
    project_id?: string;
    client_id?: string;
    media_url?: string;
  }): Promise<ScheduledPost> {
    const { data } = await apiClient.post<ScheduledPost>('/api/distribution/schedule', body);
    return data;
  },
  async queue(params?: { status?: string; project_id?: string }): Promise<ScheduledPost[]> {
    const { data } = await apiClient.get<ScheduledPost[]>('/api/distribution/queue', { params });
    return data;
  },
  async publishNow(id: string): Promise<ScheduledPost> {
    const { data } = await apiClient.post<ScheduledPost>(`/api/distribution/publish/${id}`);
    return data;
  },
};
